site/content/integrate/apps/lifecycle/callbacks.md

```diff
@@ -3,5 +3,15 @@ title: Callbacks
+|---------------------|--------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| OnInstall           | `on_install`             | OnInstall gets invoked when a System Admin installs the App with a `/apps install` command. It may return another call to the app, or a form to display.                                                                                                                                                                                                                            |
+| OnUninstall         | `on_uninstall`           | OnUninstall gets invoked when a System Admin uses the `/apps uninstall` command, before the app is actually removed.                                                                                                                                                                                                                                                                |
+| OnVersionChanged    | `on_version_changed`     | OnVersionChanged gets invoked when the Mattermost-recommended version of the app no longer matches the previously installed one, and the app needs to be upgraded/downgraded.                                                                                                                                                                                                       |
+| GetOAuth2ConnectURL | `get_oauth2_connect_url` | GetOAuth2ConnectURL is called when the App's "connect to third party" link is selected to start an OAuth flow. It must return Data set to the remote OAuth2 redirect URL. A "state" string is created by the proxy, and is passed to the app as a value. The state is a one-time secret that is included in the connect URL, and will be used to validate OAuth2 complete callback. |
+| OnOAuth2Complete    | `on_oauth2_complete`     | OnOAuth2Complete gets called upon successful completion of the remote (third party) OAuth2 flow, and after the "state" has already been validated. It gets passed the URL query as Values. The App should obtain the OAuth2 user token, and store it persistently for future use using `appclient.StoreOAuth2User`.                                                                 |
+| OnRemoteWebhook     | `on_remote_webhook`      | OnRemoteWebhook gets invoked when an HTTP webhook is received from a remote system, and is optionally authenticated by Mattermost. The request is passed to the call serialized as HTTPCallRequest (JSON).                                                                                                                                                                          |
```

Comment:
Do any of these have default values if not provided?

Reply:
I have no idea. If the callback isn't defined (i.e. doesn't exist in the manifest) then the callback is not executed. Why would the default value of these callback definitions come into play?

Reply:
I think some of them may have defaults like `on_install`. I don't know the current state of this

Reply:
> Why would the default value of these callback definitions come into play?

If an app doesn't supply an `on_install`, I think it still gets pinged on install

Reply:
@mickmister Since these are events triggered on the Mattermost side, aren't they _all_ technically pinged when the event happens, and only implemented callbacks actually exist to get called? This would support your understanding that `on_install` is always called.

Reply:
For `on_install`, it is only invoked if it is defined in the manifest: https://github.com/mattermost/mattermost-plugin-apps/blob/0856d8f8eac70040d6715f7c3733a3d3ca288ddc/server/proxy/install.go#L92